worlds/ff4fe/__init__.py

```diff
@@ -0,0 +1,488 @@
+                # ...but otherwise all underground locations require underground access.
+                add_rule(self.get_location(location.name),
+                         lambda state: state.has("Hook", self.player) or state.has("Magma Key", self.player))
+            # Moon needs the ability to get to the moon.
+            if location.area in topology.moon_areas:
+                add_rule(self.get_location(location.name),
+                         lambda state: state.has("Darkness Crystal", self.player))
+                if self.options.UnsafeKeyItemPlacement:
+                    add_rule(self.get_location(location.name),
+                             lambda state: state.has("Hook", self.player) or state.has("Magma Key", self.player))
+            # Otherwise, we consult the list of area-specific rules (e.g. Baron Castle requires Baron Key)..
+            if location.area in FERules.area_rules.keys():
+                for requirement in FERules.area_rules[location.area]:
+                    add_rule(self.get_location(location.name),
+                             lambda state, true_requirement=requirement: state.has(true_requirement, self.player))
+            # Major slots must have useful or better.
+            if location.major_slot and location.name not in self.options.exclude_locations:
+                add_item_rule(self.get_location(location.name),
+                              lambda item: (item.classification & (ItemClassification.useful | ItemClassification.progression)) > 0)
+            # The "harder" an area, the more key items and characters we need to access.
+            # This does two things. First, it ensures a wider distribution of key items so sphere 1 can't just be one
+            # piece of progression, which isn't fun in FE.
+            # Second of all, it includes characters in the requirements so a lategame FE area isn't equivalent to an early
+            # Zelda dungeon for progression balancing.
+            # Also it makes the spoiler playthrough nicer.
+            for i in range(len(FERules.location_tiers.keys())):
+                if (location.area in FERules.location_tiers[i]
+                        and self.options.ItemPlacement.current_key == "normal"
+                        and location.name not in locations.character_locations
+                        and not location.name.startswith("Objective")):
+                    add_rule(self.get_location(location.name),
+                             lambda state, tier=i: state.has_group("characters",
+                                                           self.player,
+                                                           FERules.logical_gating[tier]["characters"]) and
+                                           state.has_group("key_items",
+                                                           self.player,
+                                                           FERules.logical_gating[tier]["key_items"]))
```

Comment:
Will it create issues if someone adds a character to their `start_inventory`? Since the randomizer might see "ah, you have 2 of character A and 1 of character B, so you can get this check that requires 3 characters".
If that would be an issue, you should use `has_group_unique` instead.

Reply:
Characters are not unique, so it's correct to not use has_group_unique. While it's true adding them to start inventory will cause the generator to think there are checks that are now "in logic", this will have little effect in practice: 

- characters in start inventory don't do anything (they're items for convenience's sake on the part of options and the code, not because they're equivalent to a Life potion or whatnot), 
- the gating logic is about spreading things out so at least some progression is located in less difficult areas.
- that gating logic still requires key items as well, so nothing will change.
- and even if there were changes, it makes things slightly more difficult for the player but doesn't render the game unbeatable.